Source/DafnyCore/Compilers/SinglePassCompiler.cs

```diff
@@ -1047,9 +1049,134 @@ public class TypeArgumentInstantiation {
     protected abstract ILvalue EmitMemberSelect(Action<ConcreteSyntaxTree> obj, Type objType, MemberDecl member, List<TypeArgumentInstantiation> typeArgs, Dictionary<TypeParameter, Type> typeMap,
       Type expectedType, string/*?*/ additionalCustomParameter = null, bool internalAccess = false);
 
-    protected void EmitArraySelect(string index, Type elmtType, ConcreteSyntaxTree wr) {
-      EmitArraySelect(new List<string>() { index }, elmtType, wr);
+    protected enum MemberCompileStatus { Ordinary, Identity, AlwaysTrue }
+
+    protected MemberCompileStatus GetMemberStatus(MemberDecl member) {
+      if (member.EnclosingClass is DatatypeDecl dt) {
+        if (OptimizesInvisibleDatatypeWrappers && IsInvisibleWrapper(dt, out var dtor) && dtor == member) {
+          // "member" is the sole destructor of an invisible datatype wrapper
+          return MemberCompileStatus.Identity;
+        } else if (member is DatatypeDiscriminator) {
+          // a discriminator of an inductive or coinductive datatype
+          return dt.Ctors.Count(c => !c.IsGhost) == 1 ? MemberCompileStatus.AlwaysTrue : MemberCompileStatus.Ordinary;
+        }
+      }
+
+      return MemberCompileStatus.Ordinary;
     }
+
+    protected bool IsInvisibleWrapper(DatatypeDecl dt, out DatatypeDestructor coreDestructor) {
+      if (!DafnyOptions.O.OptimizeInvisibleDatatypeWrappers ||
+          !OptimizesInvisibleDatatypeWrappers ||
+          !(dt is IndDatatypeDecl) ||
+          dt.IsExtern(out _, out _) ||
+          dt.Members.Any(member => member is Field)) {
+        // don't optimize
+      } else {
+        var i = dt.Ctors.FindIndex(ctor => !ctor.IsGhost);
+        if (0 <= i) {
+          if (dt.Ctors.FindIndex(i + 1, ctor => !ctor.IsGhost) == -1) {
+            // there is exactly one non-ghost constructor
+            var ctor = dt.Ctors[i];
+            var j = ctor.Destructors.FindIndex(dtor => !dtor.IsGhost);
+            if (0 <= j) {
+              if (ctor.Destructors.FindIndex(j + 1, dtor => !dtor.IsGhost) == -1) {
+                // there is exactly one non-ghost parameter to "ctor"
+                var candidateCoreDestructor = ctor.Destructors[j];
```

Comment:
I thought, in my old-school brain, that the indices were clear enough for this task. Your suggested names would be much better than my `i` and `j`.

Yes, it was the creation of a new `List` that I was avoiding. Then again, I imagine the creation of the various lambda expressions (like `d => !d.IsGhost`) will take as much time and space as an additional `List`, so not much is saved. Because I used `firstDestructorIndex + 1` as the starting point, I believe the list would be traversed only once, but unless the C# compiler is good at optimizing, it would create a lambda expression for each invocation of `FindIndex`.

Regarding copying the entire AST and omitting ghost things: I'm very much in favor of this. A lot of the code that is duplicated across the compilers has to do with avoiding ghosts, so that would be so nice to do once and for all.